### PR TITLE
Handle axis_location=None correctly 

### LINKED ISSUE #4042: Generalize the configuration of axes
One of my most common use cases in the visualization is to present _an overview of some system's dynamics, plotting ~10-20 signals on the vstacked graphs (from 2 to 5 signals per plot)_. Usually I use as many extra axes as possible in order to keep the good perception ratio (up to 4 y-axes and 2 x-axes at the same time). For sure, there are other use cases, but, please, consider this as an example.

I was using `matplotlib` a lot, but it has some issues with sharing the plots for further analysis. At the same time, `bokeh` has very nice capabilities to zoom the shared plots, easily select data / data intervals, etc.

To facilitate the visual style configuration (labels, limits, scales, colors, markers, etc) I usually use either nested dict or DataFrame. This helps to keep the same visual style / axes config across the plots. This is the way it works for me:

- Create `signal -> visual style` mapping;
- Allocate grid of plots (`axis[0], axis_right[0], axis_right2[0], axis_upper[0], ..., axis[n], ...`);
- Create `subplot, axis -> signal` mapping;
- Iterate over mapping to plot the signals and apply the visual styles. 

Looking at `bokeh`, I see that the regular plotting pipeline slightly differs, but still is more or less the same. The only thing that is really missing (at least as I'm aware of), is the possibility to iterate over base x/y-axes and extra x/y-axes in the same way.

Here is the example how it works:
```
import numpy as np
import pandas as pd
 
from bokeh.io import gridplot, output_file, show
from bokeh.plotting import figure
from bokeh.models import LinearAxis, Range1d
 
output_file("test.html")
 
df = pd.DataFrame([])
df['time'] = [1, 2, 3, 4, 5]
df['temperature'] = [30, 31, 32.5, 33, 31]
df['humidity'] = [10, 30, 50, 20, 40]
df['sensor'] = [0.22, 0.12, 0.14, 0.08, 0.04]
 
s1 = figure(title=None)
s1.x_range = Range1d(start=-1, end=6)
s1.y_range = Range1d(start=25, end=35)
 
s1.extra_y_ranges = {
    'humidity': Range1d(start=0, end=50),
    'sensor': Range1d(start=0, end=0.25)
}
 
s1.line(x=df['time'], y=df['temperature'],
        color='orange', legend='temperature')
s1.line(x=df['time'], y=df['humidity'], y_range_name="humidity",
        color="blue", alpha=0.5, legend='humidity')
s1.line(x=df['time'], y=df['sensor'], y_range_name="sensor",
        color="red", legend='sensory data')
 
s1.xaxis.axis_label = 'Reference Units'
s1.yaxis.axis_label = 'Temperature'
s1.add_layout(LinearAxis(y_range_name='humidity', axis_label='Humidity'), 'left')
s1.add_layout(LinearAxis(y_range_name='sensor', axis_label='Sensor'), 'right')
 
show(s1)
```

And example how I'd appreciate it to work (just to show the concept):
```
import numpy as np
import pandas as pd
 
from bokeh.io import gridplot, output_file, show
from bokeh.plotting import figure
from bokeh.models import LinearAxis, Range1d
 
output_file("test.html")
 
df = pd.DataFrame([])
df['time'] = [1, 2, 3, 4, 5]
df['temperature'] = [30, 31, 32.5, 33, 31]
df['humidity'] = [10, 30, 50, 20, 40]
df['sensor'] = [0.22, 0.12, 0.14, 0.08, 0.04]

s1 = figure(title=None)

s1.x_ranges = {
    'reference units': Range1d(start=-1, end=6),
}
s1.y_ranges = {
    'temperature': Range1d(start=0, end=50),
    'humidity': Range1d(start=0, end=50),
    'sensor': Range1d(start=0, end=0.25)
}
 
s1.line(x=df['time'], y=df['temperature'], 
        x_range_name='reference_units', y_range_name="temperature",
        color='orange', legend='temperature')
s1.line(x=df['time'], y=df['humidity'], 
        x_range_name='reference_units', y_range_name="humidity",
        color="blue", alpha=0.5, legend='humidity')
s1.line(x=df['time'], y=df['sensor'], 
        x_range_name='reference_units', y_range_name="sensor",
        color="red", legend='sensory data')
 
s1.add_layout_x(LinearAxis(x_range_name='reference units', axis_label='Reference Units'), 'bottom')
s1.add_layout_y(LinearAxis(y_range_name='temperature', axis_label='Temperature'), 'left')
s1.add_layout_y(LinearAxis(y_range_name='humidity', axis_label='Humidity'), 'left')
s1.add_layout_y(LinearAxis(y_range_name='sensor', axis_label='Sensor'), 'right')
 
show(s1)
```

### PR DESCRIPTION
**Pull request by bryevdv**
issues: fixes #4042

### CODE REVIEW

**Comment by bryevdv** [comment]
Still want to refactor that redundant axis creation  code, and add some tests. 

Note we create the axis with `plot=None` just to get its ticker for use with grids. Would be nice to improve this somehow later.

**Comment by bryevdv** [comment]
@soupault the following code
```
import pandas as pd

from bokeh.io import gridplot, output_file, show
from bokeh.plotting import figure
from bokeh.models import LinearAxis, Range1d


output_file("test.html")

df = pd.DataFrame([])
df['time'] = [1, 2, 3, 4, 5]
df['temperature'] = [30, 31, 32.5, 33, 31]
df['humidity'] = [10, 30, 50, 20, 40]
df['sensor'] = [0.22, 0.12, 0.14, 0.08, 0.04]

s1 = figure(title=None, x_axis_location=None, y_axis_location=None)

s1.extra_x_ranges['time'] = Range1d(-1, 6)
s1.extra_y_ranges['temperature'] = Range1d(25, 35)
s1.extra_y_ranges['humidity'] = Range1d(0, 50)
s1.extra_y_ranges['sensor'] = Range1d(0, 0.25)

s1.line(x=df['time'], y=df['temperature'],
        x_range_name='time', y_range_name='temperature',
	color='orange', legend='temperature')
s1.line(x=df['time'], y=df['humidity'],
        x_range_name='time', y_range_name="humidity",
	color="blue", alpha=0.5, legend='humidity')
s1.line(x=df['time'], y=df['sensor'],
        x_range_name='time', y_range_name="sensor",
	color="red", legend='sensory data')

s1.add_layout(LinearAxis(x_range_name='time', axis_label='Time'), 'below')
s1.add_layout(LinearAxis(y_range_name='temperature', axis_label='Temperature'), 'left')
s1.add_layout(LinearAxis(y_range_name='humidity', axis_label='Humidity'), 'left')
s1.add_layout(LinearAxis(y_range_name='sensor', axis_label='Sensor'), 'right')

show(s1)
```
generates this output
screen shot 2016-03-19 at 10 21 46 pm

Not that if you also want to suppress the automatic grid creation (because they don't match up with any of the extra axes) then you'd pass e.g., `x_axis_type=None` to `figure` which also surpassed default axis creation (I overlooked to suggest earlier).

**Comment by bryevdv** [comment]
@bokeh/dev this is ready for review

**Comment by soupault** [comment]
@bryevdv Great, thanks a lot for your work and time! :cake: 
I agree, that the default grid is the best option so far.

Regarding the documentation for these arguments: where would you prefer to have it?
I remember only these two sections:
* http://bokeh.pydata.org/en/0.11.0/docs/reference/plotting.html#bokeh.plotting.figure.figure
* http://bokeh.pydata.org/en/0.11.0/docs/user_guide/plotting.html#twin-axes

**Comment by bryevdv** [comment]
@soupault I would suggest:
* adding to `plotting.figure.Figure` class docstring
* adding a new section to http://bokeh.pydata.org/en/latest/docs/user_guide/plotting.html for axis properies configurable via `Figure`

File: bokeh/plotting/helpers.py [hunk 1]
```diff
@@ -247,6 +247,30 @@ def _tool_from_string(name):
...
+    axiscls = _get_axis_class(axis_type, rng)
+    if axiscls:
+
+        if axiscls is LogAxis:
+            plot.x_mapper_type = 'log'
+
+        # this is so we can get a ticker off the axis, even if we discard it
+        axis = axiscls(plot=plot if axis_location else None)
+
+        if hasattr(axis.ticker, 'num_minor_ticks'):
```
Comment (mattpap) [inline bokeh/plotting/helpers.py]: I presume `isinstance(..., ContinuousAxis)` should be used here.

File: bokeh/plotting/helpers.py [hunk 2]
```diff
@@ -247,6 +247,30 @@ def _tool_from_string(name):
...
+            axis.ticker.num_minor_ticks = _get_num_minor_ticks(axiscls, minor_ticks)
+
+        axis_label = axis_label
+        if axis_label:
+            axis.axis_label = axis_label
+
+        grid = Grid(plot=plot, dimension=dim, ticker=axis.ticker); grid
```
Comment (mattpap) [inline bokeh/plotting/helpers.py]: `; grid` is unnecessary or is this some dirty magic (TS API raised some questions)?

Comment (bryevdv) [inline bokeh/plotting/helpers.py]: silences pyflakes

File: bokeh/plotting/helpers.py (hunk 1 above)
Comment (bryevdv) [inline bokeh/plotting/helpers.py]: Ah, yes at this point that seems possible and better (was not previously)

File: bokeh/plotting/helpers.py (hunk 2 above)
Comment (mattpap) [inline bokeh/plotting/helpers.py]: I see. So on related note, how this gets added to `fig.renderers`?

File: bokeh/plotting/helpers.py (hunk 1 above)
Comment (mattpap) [inline bokeh/plotting/helpers.py]: Yeah, I wasn't aware of this class my self, just `hasattr()`  used for this purpose makes me look for alternatives.

File: bokeh/plotting/helpers.py (hunk 2 above)
Comment (bryevdv) [inline bokeh/plotting/helpers.py]: It adds itself, if `plot` param is not `None`. We could re-visit this at some point.

Comment (mattpap) [inline bokeh/plotting/helpers.py]: I was asking from TS API perspective. Without this piece of knowledge this initialization code looks like nonsense and obviously doesn't work in javascript. OK, now I know how to proceed.

Comment (mattpap) [inline bokeh/plotting/helpers.py]: > We could re-visit this at some point.

Definitively, in another PR.

**Comment by bryevdv** [comment]
@mattpap any other comments? If not I will merge when tests pass

**Comment by bryevdv** [comment]
NM, log plot axis is broken, must fix first.

**Comment by bryevdv** [comment]
log plots fixed

File: bokeh/plotting/helpers.py [hunk 3]
```diff
@@ -247,6 +247,36 @@ def _tool_from_string(name):
...
+            else:
+                raise ValueError("received invalid dimension value: %r" % dim)
+
+        # this is so we can get a ticker off the axis, even if we discard it
+        axis = axiscls(plot=plot if axis_location else None)
+
+        if isinstance(axis.ticker, ContinuousTicker):
+            axis.ticker.num_minor_ticks = _get_num_minor_ticks(axiscls, minor_ticks)
+
+        axis_label = axis_label
```
Comment (mattpap) [inline bokeh/plotting/helpers.py]: @bryevdv, ^.

Comment (bryevdv) [inline bokeh/plotting/helpers.py]: Well, harmless at least. :) Thanks for the catch I will remove in some other PR soon

**Pull request merged**